### DISCUSSION TITLE
How will `rye` and `uv` coexist in the future?

### DISCUSSION THREAD

**Discussion by nikhilweee** [astral-sh/rye#1164] [Q&A] [answered]
I have been following the two projects and needless to say the progress so far has been great, but now that both `uv` and `rye` fall under the astral umbrella, what does the future hold for both these projects? The reason I ask this is to help new adopters like myself make an informed choice about which project to invest their time in and understand future roadmap.

**Is `uv` going to subsume `rye`**? 

When it comes to "full-service" package managers for python like poetry or pdm, rye seems to better fit the bill. However, uv was born out of the astral umbrella with the aim of achieving the goal of "cargo for python". Until recently I used to think as these two projects as being complimentary to each other. I don't claim to fully understand the two but from my limited understanding rye seemed to be the "frontend" whereas uv seemed to be the "backend" for rye. If someone needed an ultra fast python dependency manager, they could use uv. If they need to manage their project in addition to handling dependencies, they could consider rye.

That said, with the [recent addition](https://github.com/astral-sh/uv/pull/4193) of `uv add` and `uv remove`, my theory is starting to fall apart, and I am confused.  With @zanieb's recent [comment](https://github.com/astral-sh/uv/issues/1870#issuecomment-2177309018), it seems that uv is on the path to evolve into a full fledged package manager, and I think falls into the rye territory. uv is developing at a much faster pace with multiple releases happening every week, whereas rye seems to be following a monthly cadence. Does this suggest that `uv` is eventually going to subsume rye?

If that is the case, I think that future adopters of the project deserve to know. It is easy to get started with a new tool, but the nuances are hard to master. Getting your friends to try a new tool is easy, but getting your colleagues to do so is hard. If rye is going to be deprecated in favor of uv, I would rather spend my time mastering uv and wait for feature parity in the future instead of learning rye right now only to unlearn it in the future.

**Is `rye` going to coexist with `uv` in the future?**

I find it hard to imagine such a future for reasons mentioned above, but if both projects are indeed headed in this direction, it will be crucial to separate responsibilities so that the projects don't compete against each other any more than they already do.

In either case it would be great if one of the maintainers could chime in and clarify how these projects will coexist in the future.

**Comment by zanieb** [comment, marked as answer]
Hi! Charlie is on vacation right now, and I'm sure he can provide more context when he's back, but I appreciate your concerns and want to give the short answer now. 

Armin wrote a [blog post](https://lucumr.pocoo.org/2024/2/15/rye-grows-with-uv/) about this topic after the launch of uv. To quote from there:

> Rye will continue to be a test bed for what Python tooling can be... As the user experience becomes more obvious uv itself will turn from what it is today — low level plumbing — into that higher level tool with a clear migration path of folks using rye to that new uv.

This was a few months ago and we're clearly devoting a lot of resources to exploring new workflows in uv right now. I think this is mostly because it's been easier for us to build on top of uv's code rather than Rye's. We're focusing on reaching parity with tools like pipx and Poetry while adding Rye-inspired features like workspaces, Python toolchain management, and dependency sources. There are new uv workflows that overlap with Rye, but there are also Rye features that aren't a focus yet like running tests, building wheels, and publishing packages. Rye is still functioning as a place to explore innovation in Python tooling, it can move quickly by making compromises like using other tools under the hood. 

I don't see Rye being deprecated soon. The new features in uv need time to mature and we're interested in growing our role as maintainers of Rye. We don't see the projects as competitive; uv has made Rye faster and we get a lot of feedback and insights from Rye users. We want to support early adopters of both tools, and an easy migration from Rye to uv is a priority.

**Comment by nikhilweee** [comment]
> We want to support early adopters of both tools, and an easy migration from Rye to uv is a priority.

Thanks for the answer, I think this instills some confidence and takes away some hesitation from trying both these tools. Unless you wish to use this discussion to reach out to a broader audience, please feel free to close this.
